### PR TITLE
[JENKINS-48309] - Prevent TimeoutException in AsyncFutureImpl in the case of spurious wakeups

### PR DESCRIPTION
**Pull request by oleg-nenashev**
**TL;DR**: A large part of Jenkins’ get-with-timeout logic is a subject for improper exit before the timeout.

It happens, because "wait(timeout)" is not in the loop. Object#wait(long) is explicit about that: "A thread can also wake up without being notified, interrupted, or timing out, a so-called spurious wakeup. While this will rarely occur in practice, applications must guard against it by testing for the condition that should have caused the thread to be awakened, and continuing to wait if the condition is not satisfied. In other words, waits should always occur in loops..."

https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#wait-long-

Impact:

* AsyncFutureImpl is being used explicitly in JarCacheSupport. Jar resolution with a timeout may fail due to the issue
* The class is overridden in Jenkins Core's public API, e.g. hudson.model.queue.FutureImpl. It's hard to estimate risks in the core, but it looks bad.

https://issues.jenkins-ci.org/browse/JENKINS-48309

@reviewbybees @rysteboe @jglick

### CODE REVIEW

File: src/main/java/hudson/remoting/AsyncFutureImpl.java [hunk 1]
```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
...
     }
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        long timeoutMs = unit.toMillis(timeout);
+        // It is not very accurate, but all(proof?) production uses of this class pass seconds.
+        // Test cases pass 1 millisecond and other such fun, but wait(long) is not time-accurate anyway
+        long endWaitTime = System.currentTimeMillis() + timeoutMs;
```
Comment (stephenc) [inline src/main/java/hudson/remoting/AsyncFutureImpl.java]: :bug: will cause massive issues as the system time changes. You need to use nanotime instead... care for comparison logic though

File: src/main/java/hudson/remoting/AsyncFutureImpl.java [hunk 2]
```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
...
+        }
+
         if(!completed)
             throw new TimeoutException();
         if(cancelled)
             throw new CancellationException();
+
+        //TODO: we may be calling a complex get() implementation in the override, which may hang the code
+        // The only missing behavior is "problem!=null" branch, but probably we cannot just replace the code without
+        // an override issue risk.
```
Comment (stephenc) [inline src/main/java/hudson/remoting/AsyncFutureImpl.java]: Then perhaps move the get outside the synchronised

**Comment by oleg-nenashev** [comment]
Probably it causes https://issues.jenkins-ci.org/browse/JENKINS-20947, not 100% sure.

File: src/main/java/hudson/remoting/AsyncFutureImpl.java (hunk 2 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/AsyncFutureImpl.java]: I will need to investigate the implementations first. `get()` outside synchronized may help for sure, but I would like to find real examples before touching this code.

File: src/main/java/hudson/remoting/AsyncFutureImpl.java [hunk 3]
```diff
@@ -78,13 +79,28 @@ public synchronized V get() throws InterruptedException, ExecutionException {
...
+        // The accuracy of wait(long) operation is milliseconds anyway, but ok.
+        long endWaitTime = System.nanoTime() + unit.toNanos(timeout);
+        while (!completed) {
+            long timeToWait = endWaitTime - System.nanoTime();
+            if (timeout < 0) {
+                break;
+            }
+
+            long timeToWaitMs = Math.max(1, timeToWait / 1000000);
+            wait(timeToWaitMs);
```
Comment (stephenc) [inline src/main/java/hudson/remoting/AsyncFutureImpl.java]: There is `wait(long,int)` to wait with “nanosecond” precision

File: src/main/java/hudson/remoting/AsyncFutureImpl.java (hunk 2 above)
Comment (stephenc) [reply src/main/java/hudson/remoting/AsyncFutureImpl.java]: Ack

File: src/main/java/hudson/remoting/AsyncFutureImpl.java (hunk 3 above)
Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/AsyncFutureImpl.java]: The implementation in my Java does the same, actually. Decompiled code:

```
public final void wait(long timeout, int nanos) throws InterruptedException {
        if (timeout < 0) {
            throw new IllegalArgumentException("timeout value is negative");
        }

        if (nanos < 0 || nanos > 999999) {
            throw new IllegalArgumentException(
                                "nanosecond timeout value out of range");
        }

        if (nanos >= 500000 || (nanos != 0 && timeout == 0)) {
            timeout++;
        }

        wait(timeout);
    }
```

I will patch the impl though

**Comment by oleg-nenashev** [comment]
ping @stephenc

File: src/main/java/hudson/remoting/AsyncFutureImpl.java [hunk 4]
```diff
@@ -78,13 +79,27 @@ public synchronized V get() throws InterruptedException, ExecutionException {
...
-            wait(unit.toMillis(timeout));
+        // The accuracy of wait(long) operation is milliseconds anyway, but ok.
+        long endWaitTime = System.nanoTime() + unit.toNanos(timeout);
+        while (!completed) {
+            long timeToWait = endWaitTime - System.nanoTime();
+            if (timeout < 0) {
+                break;
+            }
+
+            wait(timeToWait / 1000000, (int)(timeToWait % 1000000));
```
[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/java/hudson/remoting/AsyncFutureImpl.java]: underscores in java 8  `1_000_000`

File: src/main/java/hudson/remoting/AsyncFutureImpl.java [hunk 5]
```diff
@@ -78,13 +79,27 @@ public synchronized V get() throws InterruptedException, ExecutionException {
...
 
+    @CheckForNull
     public synchronized V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        if(!completed)
-            wait(unit.toMillis(timeout));
+        // The accuracy of wait(long) operation is milliseconds anyway, but ok.
+        long endWaitTime = System.nanoTime() + unit.toNanos(timeout);
+        while (!completed) {
+            long timeToWait = endWaitTime - System.nanoTime();
+            if (timeout < 0) {
```
Comment (stephenc) [inline src/main/java/hudson/remoting/AsyncFutureImpl.java]: `s/timeout/timeToWait/`

Comment (oleg-nenashev) [reply src/main/java/hudson/remoting/AsyncFutureImpl.java]: oh yes :( It makes me think that some test coverage could be actually useful

**Comment by oleg-nenashev** [comment]
Test hanging is unrelated, it comes from the previous change. Git bisect points to `[7bd389e7991efedf0cd67afb7ce821abb50891bb] Javadoc: cleanup empty tag Javadoc warnings` after 2 attempts, looks promising 🤷‍♂️
